query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+                *entry.get() == this_type,
+            // If not, record the one we just determined.
+            Entry::Vacant(entry) => {
+                entry.insert(this_type);
+                true
+            }
+        }
+    }
+
+    /// Ensure that the given place has the correct types to be a tx-id.
+    /// Right now this is mostly unimplemented: we fail hard if anything but a placeholder is
+    /// present.
+    fn mark_as_tx(&mut self, tx: &PatternNonValuePlace) {
+        match *tx {
+            PatternNonValuePlace::Placeholder => (),
+            _ => unimplemented!(),           // TODO
+        }
+    }
+
+    /// Ensure that the given place can be an entity, and is congruent with existing types.
+    /// This is used for `entity` and `attribute` places in a pattern.
+    fn mark_as_ref(&mut self, value: &PatternNonValuePlace) {
+        // If it's a variable, record that it has the right type.
+        // Ident or attribute resolution errors (the only other check we need to do) will be done
+        // by the caller.
+        if let &PatternNonValuePlace::Variable(ref v) = value {
+            if !self.constrain_var_to_type(v.clone(), ValueType::Ref) {
+                self.mark_known_empty("Couldn't constrain var to Ref.");
+            }
+        }
+    }
+
+    fn mark_known_empty(&mut self, why: &str) {
+        self.is_known_empty = true;
+        println!("{}", why);                   // TODO: proper logging.
+    }
+
+    fn entid_for_ident<'s, 'a>(&self, schema: &'s Schema, ident: &'a NamespacedKeyword) -> Option<i64> {
+        schema.get_entid(&ident.to_string())      // TODO: #291.
+    }
+
+    fn attribute_for_ident<'s, 'a>(&self, schema: &'s Schema, ident: &'a NamespacedKeyword) -> Option<&'s Attribute> {
+        self.entid_for_ident(schema, ident).and_then(|x| schema.attribute_for_entid(x))
+    }
+
+    fn table_for_attribute_and_value<'s, 'a>(&self, attribute: &'s Attribute, value: &'a PatternValuePlace) -> Option<DatomsTable> {
+        if attribute.fulltext {
+            match value {
+                &PatternValuePlace::Placeholder =>
+                    Some(DatomsTable::Datoms),            // We don't need the value.
+
+                // TODO: an existing non-string binding can cause this pattern to fail.
+                &PatternValuePlace::Variable(_) =>
+                    Some(DatomsTable::AllDatoms),
+
+                &PatternValuePlace::Constant(NonIntegerConstant::Text(_)) =>
+                    Some(DatomsTable::AllDatoms),
+
+                _ => {
+                    // We can't succeed if there's a non-string constant value for a fulltext
+                    // field.
+                    None
+                }
+            }
+        } else {
+            Some(DatomsTable::Datoms)
+        }
+    }
+
+    fn table_for_places<'s, 'a>(&self, schema: &'s Schema, attribute: &'a PatternNonValuePlace, value: &'a PatternValuePlace) -> Option<DatomsTable> {
```

Comment:
Throughout this implementation, little snippets of EDN would make this easier to understand.  Things like, "If we have `[... _ v]`, then ... but if we have `[... ?binding v]` then...".  It can be tricky to build those things out of the code and descriptions.